i18n-helpers/src/bin/mdbook-xgettext.rs

```diff
@@ -86,21 +103,161 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
+    let summary_template_destination = if depth == 0 {
+        PathBuf::from("messages")
+    } else if depth == 1 {
+        PathBuf::from("summary")
+    } else {
+        PathBuf::from(&current_top_level).join("summary")
+    };
+    catalogs.insert(summary_template_destination, catalog);
+
     // Next, we add the chapter contents.
-    for item in ctx.book.iter() {
-        if let BookItem::Chapter(chapter) = item {
-            let path = match &chapter.path {
+    for item in &ctx.book.sections {
+        if let BookItem::PartTitle(title) = item {
+            // Iterating through the book in section-order, the PartTitle
+            // represents the 'section' that each chapter exists within.
+            current_top_level = slug(title);
+        } else if let BookItem::Chapter(chapter) = item {
+            let source = match &chapter.path {
                 Some(path) => ctx.config.book.src.join(path),
                 None => continue,
             };
+
+            let template_destination = if depth == 0 {
+                PathBuf::from("messages")
+            } else if depth == 1 {
+                PathBuf::from(current_top_level.clone())
+            } else {
+                // The current chapter is already at depth 2, so append the chapter's
+                // name for depths greater than 1.
+                PathBuf::from(current_top_level.clone()).join(slug(&chapter.name))
+            };
+
+            // Add the (destination, catalog) to the map if it doesn't yet
+            // exist, so messages can be appended to the catalog.
+            if let None = catalogs.get(&template_destination) {
+                catalogs.insert(
+                    template_destination.clone(),
+                    Catalog::new(generate_catalog_metadata(&ctx)),
+                );
+            }
+            let catalog: &mut Catalog = catalogs.get_mut(&template_destination).unwrap();
+
             for (lineno, msgid) in extract_messages(&chapter.content) {
-                let source = format!("{}:{}", path.display(), lineno);
-                add_message(&mut catalog, &msgid, &source);
+                let source_with_lineno = format!("{}:{}", source.display(), lineno);
+                add_message(catalog, &msgid, &source_with_lineno);
+            }
+
+            // Add the contents for all of the sub-chapters within the current chapter.
+            for ChapterTemplateInfo {
+                content,
+                source,
+                template_destination,
+            } in get_subcontent_for_chapter(chapter, template_destination, depth, 2)
+            {
+                if let None = catalogs.get(&template_destination) {
+                    catalogs.insert(
+                        template_destination.clone(),
+                        Catalog::new(generate_catalog_metadata(&ctx)),
+                    );
+                }
+                let catalog = catalogs.get_mut(&template_destination).unwrap();
```

Comment:
This is a great example of a place where the [entry API](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#examples) can be used. You should be able to use `catalogs.entry(template_destination).or_insert_with(|| ...)` instead.

Two advantages of this:

- You avoid hashing the key twice
- You avoid the `unwrap()` call at the end — unwraps are always scary since they are places where the program can crash.

Reply:
Thanks! I know about the entry API but still have a lot to learn about the special functions and their uses. This shortens this code block by a bunch